testpar/t_subfiling_vfd.c

```diff
@@ -0,0 +1,260 @@
+ */
+static hid_t
+create_subfiling_ioc_fapl(void)
+{
+    H5FD_subfiling_config_t *subfiling_conf = NULL;
+    H5FD_ioc_config_t *      ioc_conf       = NULL;
+    hid_t                    ioc_fapl       = H5I_INVALID_HID;
+    hid_t                    ret_value      = H5I_INVALID_HID;
+
+    if (NULL == (subfiling_conf = HDcalloc(1, sizeof(*subfiling_conf))))
+        TEST_ERROR;
+
+    if ((ioc_fapl = H5Pcreate(H5P_FILE_ACCESS)) < 0)
+        TEST_ERROR;
+
+    if ((ret_value = H5Pcreate(H5P_FILE_ACCESS)) < 0)
+        TEST_ERROR;
+
+    if (H5Pset_mpi_params(ret_value, comm, info) < 0)
+        TEST_ERROR;
+
+    /* Get defaults for Subfiling configuration */
+    if (H5Pget_fapl_subfiling(ret_value, subfiling_conf) < 0)
+        TEST_ERROR;
+
+    if (subfiling_conf->require_ioc) {
+        if (NULL == (ioc_conf = HDcalloc(1, sizeof(*ioc_conf))))
+            TEST_ERROR;
+
+        /* Get IOC VFD defaults */
+        if (H5Pget_fapl_ioc(ioc_fapl, ioc_conf) < 0)
+            TEST_ERROR;
+
+        if (H5Pset_mpi_params(ioc_fapl, comm, info) < 0)
+            TEST_ERROR;
+
+        if (H5Pset_fapl_ioc(ioc_fapl, ioc_conf) < 0)
+            TEST_ERROR;
+    }
+    else {
+        if (H5Pset_fapl_sec2(ioc_fapl) < 0)
+            TEST_ERROR;
+    }
+
+    subfiling_conf->ioc_fapl_id = ioc_fapl;
+
+    if (H5Pset_fapl_subfiling(ret_value, subfiling_conf) < 0)
+        TEST_ERROR;
+
+    HDfree(ioc_conf);
+    HDfree(subfiling_conf);
+
+    return ret_value;
+
+error:
+    HDfree(ioc_conf);
+    HDfree(subfiling_conf);
+
+    if ((H5I_INVALID_HID != ioc_fapl) && (H5Pclose(ioc_fapl) < 0)) {
+        H5_FAILED();
+        AT();
+    }
+    if ((H5I_INVALID_HID != ret_value) && (H5Pclose(ret_value) < 0)) {
+        H5_FAILED();
+        AT();
+    }
+
+    return H5I_INVALID_HID;
+}
+
+/*
+ * A simple test that creates and closes a file with the
+ * subfiling VFD
+ */
+static void
+test_create_and_close(void)
+{
+    H5FD_subfiling_config_t subfiling_config;
+    const char *            test_filenames[2];
+    hid_t                   file_id = H5I_INVALID_HID;
+    hid_t                   fapl_id = H5I_INVALID_HID;
+
+    if (MAINPROCESS)
+        TESTING("File creation and immediate close");
+
+    fapl_id = create_subfiling_ioc_fapl();
+    VRFY((fapl_id >= 0), "FAPL creation succeeded");
+
+    VRFY((H5Pget_fapl_subfiling(fapl_id, &subfiling_config) >= 0), "H5Pget_fapl_subfiling succeeded");
+
+    file_id = H5Fcreate("basic_create.h5", H5F_ACC_TRUNC, H5P_DEFAULT, fapl_id);
+    VRFY((file_id >= 0), "H5Fcreate succeeded");
+
+    VRFY((H5Fclose(file_id) >= 0), "File close succeeded");
+
+    test_filenames[0] = "basic_create.h5";
+    test_filenames[1] = NULL;
+    h5_clean_files(test_filenames, fapl_id);
+
+    if (H5P_DEFAULT != subfiling_config.ioc_fapl_id)
+        VRFY((H5Pclose(subfiling_config.ioc_fapl_id) >= 0), "FAPL close succeeded");
+
+    return;
+}
+
+int
+main(int argc, char **argv)
```

Comment:
Unfortunately, the current regression testing for subfiling mostly consists of a simple file create and close test in this file, as well as some selection I/O testing in testpar/t_vfd.c and the serial HDF5 tests that can be run with `make check-vfd` or `ctest` when the CMake VFD testing option is enabled. The plan is to flesh out more specific regression testing for subfiling, but development work on the feature has been the highest priority so far and the feature has mostly been tested with external applications like IOR and HACC.